
In this tutorial, we annotate cell type for single cell RNA-seq based on bulk RNA-seq reference data.

## loading data


```R
require(Lightbulb)
```

##### load Seurat object
the single cell RNA-seq data should be stored in a Seurat object, with @dr$tsne, @ident and @meta.data filled. For generating your own Seurat object, please refere to the Seurat mannual (https://satijalab.org/seurat/get_started.html)


```R
Exp_Seurat=readRDS("Lightbulb_example_ExpSeurat.rds")
```

##### load super cell data
We use super-cell matrix generated by lightbulb's **Super_cell_creation**, please refer to the super cell creation example for details (https://github.com/Arthurhe/Lightbulb/blob/master/Examples/Lightbulb_example_super_cell_creation_V1.ipynb). You can use other super-cell matrix for input, but make sure that the coloumn of genes are in the same order as bulk samples, and row names of the matrix indicates the cell position in Seurat object.


```R
super_cell_matrix=readRDS(paste0("Lightbulb_example_supercell.rds"))
super_cell_matrix[1:5,1:5]
```


<table>
<thead><tr><th></th><th scope=col>1110002L01Rik</th><th scope=col>1190002N15Rik</th><th scope=col>1500009L16Rik</th><th scope=col>1700066B19Rik</th><th scope=col>1810058I24Rik</th></tr></thead>
<tbody>
	<tr><th scope=row>855</th><td>3.807040</td><td>2.194329</td><td>2.907965</td><td>0.000000</td><td>6.127443</td></tr>
	<tr><th scope=row>612</th><td>0.000000</td><td>2.194329</td><td>0.000000</td><td>0.000000</td><td>5.716956</td></tr>
	<tr><th scope=row>1849</th><td>3.096447</td><td>2.194329</td><td>0.000000</td><td>0.000000</td><td>6.807217</td></tr>
	<tr><th scope=row>629</th><td>0.000000</td><td>2.194329</td><td>2.907965</td><td>0.000000</td><td>6.660153</td></tr>
	<tr><th scope=row>1800</th><td>0.000000</td><td>3.474725</td><td>0.000000</td><td>3.696563</td><td>6.253754</td></tr>
</tbody>
</table>



##### load bulk data
bulk RNA-seq data in table form


```R
LLTE_Tcm_Tem=readRDS("Lightbulb_example_d20_Tcell_bulk.rds")
TE_MP=readRDS("Lightbulb_example_d7_Tcell_bulk.rds")
```


```R
#D30 bulk sample
LLTE_Tcm_Tem[1:5,]
```


<table>
<thead><tr><th></th><th scope=col>NAME</th><th scope=col>Sp.LLTE#1</th><th scope=col>Sp.LLTE#2</th><th scope=col>Sp.TEM#1</th><th scope=col>Sp.TEM#2</th><th scope=col>Sp.TCM#1</th><th scope=col>Sp.TCM#2</th></tr></thead>
<tbody>
	<tr><th scope=row>17</th><td>1110002L01Rik</td><td> 15.10543    </td><td> 13.48353    </td><td> 12.23527    </td><td> 17.83553    </td><td> 22.822690   </td><td> 14.491386   </td></tr>
	<tr><th scope=row>45</th><td>1190002N15Rik</td><td> 12.03903    </td><td> 26.50983    </td><td> 39.89131    </td><td> 34.67106    </td><td> 28.278363   </td><td> 22.302188   </td></tr>
	<tr><th scope=row>55</th><td>1500009L16Rik</td><td> 18.17182    </td><td> 13.48353    </td><td> 12.23527    </td><td> 20.76345    </td><td>  6.455673   </td><td>  8.810802   </td></tr>
	<tr><th scope=row>287</th><td>1700066B19Rik</td><td> 32.27725    </td><td> 15.11182    </td><td> 19.14928    </td><td> 22.95939    </td><td> 25.550526   </td><td> 17.331678   </td></tr>
	<tr><th scope=row>407</th><td>1810058I24Rik</td><td>415.57691    </td><td>401.55861    </td><td>363.98557    </td><td>406.51666    </td><td>339.933657   </td><td>370.237927   </td></tr>
</tbody>
</table>




```R
#D7 bulk sample
TE_MP[1:5,]
```


<table>
<thead><tr><th></th><th scope=col>DESCRIPTION</th><th scope=col>D7_TE</th><th scope=col>D7_MP</th></tr></thead>
<tbody>
	<tr><th scope=row>9</th><td>Rb1cc1   </td><td> 21.90129</td><td> 21.64276</td></tr>
	<tr><th scope=row>21</th><td>Cspp1    </td><td> 11.72776</td><td> 12.16485</td></tr>
	<tr><th scope=row>22</th><td>Arfgef1  </td><td> 20.60308</td><td> 20.29343</td></tr>
	<tr><th scope=row>28</th><td>Ncoa2    </td><td> 27.08860</td><td> 23.19000</td></tr>
	<tr><th scope=row>29</th><td>Tram1    </td><td>541.67319</td><td>472.38567</td></tr>
</tbody>
</table>



##### get genes to use for annotation


```R
# get the highly variable gene from single cells (need to run FindVariableGenes on Seurat object first)
HVG=head(rownames(Exp_Seurat@hvg.info), 1500)

# which gene to use
candidate_g=Reduce(intersect, list(LLTE_Tcm_Tem$NAME,TE_MP$DESCRIPTION))
common_gene=intersect(candidate_g,HVG)
```

## prepare input
  
bulk_list is a 2 level list:  
  
bulk_list  
---sample_set1  
------sample1.1=data.frame (each row is one replicate, each column is a gene)  
------sample1.2=vector (if no replicate)  
---sample_set2  
------sample2.1=data.frame  
------sample2.2=data.frame  
------sample2.3=data.frame  
  
all the sample dataframe or vector should have the same gene order


```R
bulk_list=list(Spleen_D7=list(TE=TE_MP$D7_TE[match(common_gene,TE_MP$DESCRIPTION)],
                              MP=TE_MP$D7_MP[match(common_gene,TE_MP$DESCRIPTION)]),
               Spleen_D30=list(LLE=t(LLTE_Tcm_Tem[match(common_gene,LLTE_Tcm_Tem$NAME),2:3]),
                               TEM=t(LLTE_Tcm_Tem[match(common_gene,LLTE_Tcm_Tem$NAME),4:5]),
                               TCM=t(LLTE_Tcm_Tem[match(common_gene,LLTE_Tcm_Tem$NAME),6:7])))

#normalize bulk list, and format the data to right structure
bulk_list=bulk_list_normalize(bulk_list)
```


```R
#We are only interested in the cells from spleen in the single cell data. Select cells for mapping in Seurat object
tag_cell_for_all_cells=which(Exp_Seurat@meta.data$tissue == "Spleen")

#Select super cells that have the spleen label
tag_cell=as.numeric(rownames(super_cell_matrix)) %in% tag_cell_for_all_cells

#create the super cell matrix
sc_mat=super_cell_matrix[tag_cell,common_gene]
```

### data processing and general correlation


```R
dataset=data_preprocessing(bulk_list,sc_mat,common_gene)
```

Test whether the bulk sample set classification make sense. In the following plot, the color bar on the top and left indicate the bulk sample sets. Bulk samples shoule be highly similar within one set.


```R
options(repr.plot.width=6, repr.plot.height=6)
plot_bulk_bulk_cor(dataset)
```


![png](Lightbulb_example_annotate_cell_V0.2_files/Lightbulb_example_annotate_cell_V0.2_19_0.png)


We need to know how similar are the super cells compare to bulk sets. We can plot the correlation distribution by **plot_SC_sampleSet_cor_distribution**, and set threshold to pick the right most peak


```R
options(repr.plot.width=6, repr.plot.height=3)
par(mfrow=c(1,2))
plot_SC_sampleSet_cor_distribution(dataset)
```


![png](Lightbulb_example_annotate_cell_V0.2_files/Lightbulb_example_annotate_cell_V0.2_21_0.png)


it seems that 0.75 is a good threshold for both bulk sets, because the right most peak of both distribution start at around 0.7~0.75.


```R
threshold=c(0.75,0.75)
```

To futhur illustrate how the correlation distribute, we can plot the correlation on TSNE by using **plot_SC_sampleSet_cor_TSNE**. By setting the *thresholds* parameter, all cells fall below thresholds in correlation will be shown as gray


```R
options(repr.plot.width=9, repr.plot.height=4.5)
par(mfrow=c(1,2))
plot_SC_sampleSet_cor_TSNE(dataset,Exp_Seurat,thresholds = threshold,color_contrast=1.5)
```


![png](Lightbulb_example_annotate_cell_V0.2_files/Lightbulb_example_annotate_cell_V0.2_25_0.png)


### 2nd correlation
With the *threshold* picked, we can run the 2nd round of correlation to identify the cell types.


```R
dataset=secondary_cor(dataset,threshold)
```

We can plot the annotation on TSNE by **running plot_group_annotation_TSNE**


```R
options(repr.plot.width=9, repr.plot.height=4.5)
par(mfrow=c(1,2))
plot_group_annotation_TSNE(dataset,Exp_Seurat)
```


![png](Lightbulb_example_annotate_cell_V0.2_files/Lightbulb_example_annotate_cell_V0.2_29_0.png)


We can also plot the score for each sample in their corresponding sample set with **plot_SC_sampleSet_cor_2nd_TSNE**


```R
options(repr.plot.width=12, repr.plot.height=4)
par(mfrow=c(1,3))
plot_SC_sampleSet_cor_2nd_TSNE(dataset,Exp_Seurat,plot_sets=1)
par(mfrow=c(1,3))
plot_SC_sampleSet_cor_2nd_TSNE(dataset,Exp_Seurat,plot_sets=2)
```


![png](Lightbulb_example_annotate_cell_V0.2_files/Lightbulb_example_annotate_cell_V0.2_31_0.png)



![png](Lightbulb_example_annotate_cell_V0.2_files/Lightbulb_example_annotate_cell_V0.2_31_1.png)


### match bulk annotation with scRNA groups (across time)
We can furthur conduct cross analysis between annotated cells and the clustering result (stored in Exp_Seurat@ident) and other metadata (stored in Exp_Seurat@meta.data)  
Because we were using the super-cells for annotation, to annotate the single cells we need to label them according to clusters. We can use **metadata_reprocessing** to get the cells that's relevant for annotation, for example, all the spleen cells in this data set. This function will also store the metadata information into the *dataset* object.


```R
dataset=metadata_reprocessing(dataset,Exp_Seurat,relavant_cells = which(Exp_Seurat@meta.data$tissue == "Spleen"))
```

With metadata preocessing done. We can annoate the single cells with the bulk annotation by using **dataset**.  
We will annotate an ident cluster by looking at what percentage of the cells in given cluster have been annotated by what bulk samples. If more than *min_percent* portion of a cluster is similar to a bulk sample, then we associate the name of that bulk sample with the cluster name.


```R
dataset=celltype_annotation_by_cluster(dataset,min_percent = 0.3)
```


```R
summary(dataset)
```


                         Length Class      Mode     
    bulk_list               2   -none-     list     
    bulk_list_centered      2   -none-     list     
    bulk_df                 2   -none-     list     
    bulk_df_centered        2   -none-     list     
    sample_set_mean      3000   -none-     numeric  
    sc_mat               1500   data.frame list     
    SC_sampleSet_cor     8930   -none-     numeric  
    bulk_bulk_cor          64   -none-     numeric  
    replicate_df            2   -none-     list     
    SC_sampleSet_cor_2nd    2   -none-     list     
    celltype_assignment     2   data.frame list     
    sc_metadata            11   data.frame list     
    cluster_annotation     23   -none-     character


*dataset* object explained:<br>
*dataset\$bulk_xx* & *dataset\$replicate\_df* stored the bulk RNA-seq data <br>
*dataset\$sample_set_mean* stored the mean expression of each sample set <br>
*dataset\$sc_mat* stored the super-cell matrix <br>
*dataset\$SC_sampleSet_cor* stored the 1st super-cell sample set correlation <br>
*dataset\$bulk_bulk_cor* stored the correlation between bulk samples <br>
*dataset\$sc_metadata* stored the metadata for super-cell <br>
*dataset\$SC_sampleSet_cor_2nd* stored the 2nd super-cell bulk sample correltion within each sample set <br>
*dataset\$celltype_assignment* stored the cell type assignment for each super cell <br>
*dataset\$cluster_annotation* stored the new bulk annotated name for each cluster in Seurat@ident <br> 

### some additional plotting


```R
#options(repr.plot.width=12, repr.plot.height=6)
par(mfrow=c(1,3))
identMatch=names(dataset$cluster_annotation)
names(identMatch)=names(dataset$cluster_annotation)
plot_scMeta_NotOnlySuperCell_byIdent(dataset,Exp_Seurat,identMatch = identMatch,main="cluster id",addtext=T)

identMatch=paste0(names(dataset$cluster_annotation),":",dataset$cluster_annotation)
names(identMatch)=names(dataset$cluster_annotation)
plot_scMeta_NotOnlySuperCell_byIdent(dataset,Exp_Seurat,identMatch = identMatch,main="cluster id",addtext=F)

plot_scMeta_NotOnlySuperCell_byIdent(dataset,Exp_Seurat,main="annotated cluster",addtext=T)
```


![png](Lightbulb_example_annotate_cell_V0.2_files/Lightbulb_example_annotate_cell_V0.2_39_0.png)



```R
par(mfrow=c(1,3))
identMatch=paste0("cluster",names(dataset$cluster_annotation),": ",dataset$cluster_annotation)
names(identMatch)=names(dataset$cluster_annotation)
plot_scMeta_NotOnlySuperCell_by_category(dataset,Exp_Seurat,identMatch)
```


![png](Lightbulb_example_annotate_cell_V0.2_files/Lightbulb_example_annotate_cell_V0.2_40_0.png)



![png](Lightbulb_example_annotate_cell_V0.2_files/Lightbulb_example_annotate_cell_V0.2_40_1.png)



![png](Lightbulb_example_annotate_cell_V0.2_files/Lightbulb_example_annotate_cell_V0.2_40_2.png)



![png](Lightbulb_example_annotate_cell_V0.2_files/Lightbulb_example_annotate_cell_V0.2_40_3.png)





```R

```
